x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/logging/Throttler.java

```diff
@@ -7,152 +7,222 @@
+
+        private final AtomicLong skippedLogCalls = new AtomicLong(INITIAL_LOG_COUNTER_VALUE);
+        private final AtomicReference<Instant> timeOfLastLogCall;
         private final Clock clock;
-        private final Consumer<String> consumer;
+        private final Logger throttledLogger;
+        private final Level level;
+        private final String originalMessage;
+        private final Throwable throwable;
 
-        LogExecutor(Clock clock, Consumer<String> throttledConsumer) {
-            this(clock, 0, throttledConsumer);
-        }
-
-        LogExecutor(Clock clock, long skippedLogCalls, Consumer<String> consumer) {
-            this.skippedLogCalls = skippedLogCalls;
+        LogExecutor(Clock clock, Logger logger, Level level, String originalMessage, @Nullable Throwable throwable) {
             this.clock = Objects.requireNonNull(clock);
-            timeOfLastLogCall = Instant.now(this.clock);
-            this.consumer = Objects.requireNonNull(consumer);
+            timeOfLastLogCall = new AtomicReference<>(Instant.now(this.clock));
+            this.throttledLogger = Objects.requireNonNull(logger);
+            this.level = Objects.requireNonNull(level);
+            this.originalMessage = Objects.requireNonNull(originalMessage);
+            this.throwable = throwable;
         }
 
-        void log(String message) {
-            this.consumer.accept(message);
+        void logRepeatedMessages() {
+            var numSkippedLogCalls = skippedLogCalls.get();
+            if (hasRepeatedLogsToEmit(numSkippedLogCalls) == false) {
+                return;
+            }
+
+            String enrichedMessage;
+            if (numSkippedLogCalls == 1) {
+                enrichedMessage = Strings.format("%s, repeated 1 time, last message at [%s]", originalMessage, timeOfLastLogCall.get());
+            } else {
+                enrichedMessage = Strings.format(
+                    "%s, repeated %s times, last message at [%s]",
+                    originalMessage,
+                    skippedLogCalls,
+                    timeOfLastLogCall.get()
+                );
+            }
+
+            log(enrichedMessage);
         }
 
-        LogExecutor compute(Consumer<String> executor, Duration durationToWait) {
-            if (hasDurationExpired(durationToWait)) {
-                String messageToAppend = "";
-                if (this.skippedLogCalls == 1) {
-                    messageToAppend = ", repeated 1 time";
-                } else if (this.skippedLogCalls > 1) {
-                    messageToAppend = format(", repeated %s times", this.skippedLogCalls);
-                }
-
-                final String stringToAppend = messageToAppend;
-                return new LogExecutor(this.clock, 0, (message) -> executor.accept(message.concat(stringToAppend)));
+        private void log(String enrichedMessage) {
+            LogBuilder builder = throttledLogger.atLevel(level);
```

Comment:
I originally tried constructing the `LogBuilder` in the manager class but it will fail to log because the `LogBuilder` `log()` method must be called on the constructing thread.